Configurable Column Widths

jwparkau opened the pull request:
I have added configurable column widths for the media library and playlist editor screens.
This is done through setting a ratio in the configuration file.
The defaults are at sensible values (i.e. the current values), which are 1:1:1, 1:1, and 1:2, for media library 3 columns, media library 2 columns, and playlist editor respectively.

arybczak commented:
Thanks!

Sorry for a delay in getting into this.

It looks good, but you should parse the config into a vector when processing configuration, not every time a screen is resized.

jwparkau commented:
I think I've got it working.
It doesn't compile (some error in `mpdpp.h`) unless I merge `upstream/master` but I didn't want to push that commit because I don't know if github will make the PR messy or not.
Anyway it does compile if everything is up to date.

arybczak commented on src/screens/media_library.cpp:
```diff
@@ -193,11 +193,17 @@ MediaLibrary::MediaLibrary()
-	itsLeftColWidth = COLS/3-1;
-	itsMiddleColWidth = COLS/3;
+
+	std::vector<int> ratio_vector = Config.media_library_column_width_ratio_three;
```
Why do you copy the vector here? Same below.

arybczak commented on src/utility/option_parser.cpp:
```diff
@@ -45,6 +46,63 @@ bool yes_no(const std::string &v)
+	return true;
+}
+
+std::vector<int> check_ratio_two(const std::string &v)
```
Instead of all of these helpers I think you can generalize `list_of` to take a separator and optional length of a list, then use boost::bind to partially apply it.

arybczak commented on src/settings.h:
```diff
@@ -85,6 +85,10 @@ struct Configuration
 
 	std::string pattern;
 
+	std::vector<int> playlist_editor_column_width_ratio;
```
`int` is inaccurate, we expect positive numbers here.

jwparkau commented:
Thanks for the feedback.
1. Not sure why I copied the vector, but its fixed now.
2. I changed `int` to `size_t` to match the type of `itsLeftColStartX` and friends.
3. I removed most of the fluff helpers and put faith into the `lexical_cast`. I'm not sure what you meant by `list_of`, is it a built-in function? For now I changed it to split the string on the ":" character, try to cast each substring to `size_t`, and then check the resultant vector to make sure the length matches.
4. I used `std::bind` because I couldn't get `_1` to work with `boost::bind` for some reason.

arybczak commented:
`list_of` is in `option_parser.h`.

jwparkau commented:
I'm fairly certain that the default escape, separator, and quote chars are correct (`\`, `,`, and `"`) from reading the boost docs.
I chose not to overload `parse_ratio` because the adding the numbers and checking that the total isn't `0` part didn't seem generalizable.

arybczak commented on src/utility/option_parser.h:
```diff
@@ -56,17 +56,26 @@ DestT verbose_lexical_cast(const std::string &v)
 {
 	std::vector<ValueT> result;
-	boost::tokenizer<boost::escaped_list_separator<char>> elems(v);
+	boost::escaped_list_separator<char> ecq(e, c, q);
```
Can these be more descriptive? What is `e`, `c` and `q`?

arybczak commented on src/utility/option_parser.cpp:
```diff
@@ -45,6 +45,23 @@ bool yes_no(const std::string &v)
+	std::string e("");
+	std::string c(":");
+	std::string q("");
+	std::vector<size_t> ret = list_of<size_t>(v, verbose_lexical_cast<size_t>, length, e, c, q);
```
No need to bind e,c,q if you're not using it later.

arybczak commented:
Thanks!
